java/json/src/main/java/org/bondlib/SimpleJsonReader.java

```diff
@@ -0,0 +1,480 @@
+        final JsonToken currentToken = this.advanceParserToNextToken();
+        if (currentToken != JsonToken.VALUE_NUMBER_INT) {
+            throw createInvalidBondDataException(
+                    "Current JSON token is '%s' where '%s' for Bond 'int16' data type is expected.",
+                    currentToken,
+                    JsonToken.VALUE_NUMBER_INT);
+        }
+        // an int16 value must be represented by INT (Jackson's minimal "optimal type") and be within bounds
+        final JsonParser.NumberType numberType = this.parser.getNumberType();
+        if (numberType == JsonParser.NumberType.INT) {
+            final int valueAsInt = this.parser.getIntValue();
+            if (valueAsInt >= Short.MIN_VALUE && valueAsInt <= Short.MAX_VALUE) {
+                return (short) valueAsInt;
+            }
+        }
+        throw createInvalidBondDataException(
+                "Current JSON integer value '%s' is outside of bounds for the expected Bond data type 'int16'.",
+                this.parser.getValueAsString());
+    }
+
+    @Override
+    public final int readInt32() throws IOException {
+        final JsonToken currentToken = this.advanceParserToNextToken();
+        if (currentToken != JsonToken.VALUE_NUMBER_INT) {
+            throw createInvalidBondDataException(
+                    "Current JSON token is '%s' where '%s' for Bond 'int32' data type is expected.",
+                    currentToken,
+                    JsonToken.VALUE_NUMBER_INT);
+        }
+        // an int32 value must be represented by INT (Jackson's minimal "optimal type")
+        final JsonParser.NumberType numberType = this.parser.getNumberType();
+        if (numberType == JsonParser.NumberType.INT) {
+            final int valueAsInt = this.parser.getIntValue();
+            return valueAsInt;
+        }
+        throw createInvalidBondDataException(
+                "Current JSON integer value '%s' is outside of bounds for the expected Bond data type 'int32'.",
+                this.parser.getValueAsString());
+    }
+
+    @Override
+    public final long readInt64() throws IOException {
+        final JsonToken currentToken = this.advanceParserToNextToken();
+        if (currentToken != JsonToken.VALUE_NUMBER_INT) {
+            throw createInvalidBondDataException(
+                    "Current JSON token is '%s' where '%s' for Bond 'int64' data type is expected.",
+                    currentToken,
+                    JsonToken.VALUE_NUMBER_INT);
+        }
+        // an int64 value must be represented by either LONG or INT (Jackson's "optimal types" for the value)
+        final JsonParser.NumberType numberType = this.parser.getNumberType();
+        if (numberType == JsonParser.NumberType.LONG || numberType == JsonParser.NumberType.INT) {
+            final long valueAsLong = this.parser.getLongValue();
+            return valueAsLong;
+        }
+        throw createInvalidBondDataException(
```

Comment:
"Outside bounds" doesn't seem like an accurate description of what's happened here.

Reply:
That is the accurate description. This happens when Jackson decided that the number is too large for INT or LONG and hence number type is BIG_INT.

---
In reply to: [143861047](https://github.com/Microsoft/bond/pull/632#discussion_r143861047) [](ancestors = 143861047)

Reply:
Ah, that wasn't at all clear to me. Please comment these throws with the conditions that would cause them (e.g., `// numberType == BIG_INT`